Modernize

tamird opened the pull request:
This bumps the version to 0.1.19 because the type of `AbortController`
is not backwards compatible.

- Remove obsolete dependencies.
  - abort-controller; `AbortController` is stable since Node 15.4.0.
  - source-map-support; `--enable-source-maps` is supported since Node
    12.12.0.
- Update dependencies.
- Enable strict mode in `tsconfig.json`.
- Use `which(..., {nothrow: true})` and distinguish between
  binary-not-in-PATH and binary-access-failed. Log an error.
- Check for presence of clangd binary in zip archive before extracting.
- Use native promises APIs instead of `promisify`.
- Assign severity to all logs (no more `console.log`).

The debugging changes here are what motivated the work; I'm having some problems using vscode-clangd on a remote machine where the path to clangd is absolute, and the current implementation doesn't log anything actionable.

tamird commented:
@HighCommander4 would you be willing to review this?

HighCommander4 commented:
My preference is for changes to a piece of code to be reviewed by the original author of that code, since that person has more familiarity with the code and the reasons it was doing things a certain way.

In this case the original author of most of `src/index.ts` is @sam-mccall, and accordingly I've added him as a reviewer.

If you don't get a response in a few weeks, please feel free to flag me as a fallback.

tamird commented:
Thank you. In the meantime, could you hit the CI once more? I've replaced `clang-format` with `prettier` and fixed the formatting (along with a minimal configuration that hews as closely to the current formatting as possible).

tamird commented:
@HighCommander4 could you take a look at this please? It has been 2 weeks now.

tamird commented:
@HighCommander4 3 more weeks elapsed. Could you please give this a review?

HighCommander4 commented:
@tamird I'm sorry, I'm a volunteer contributor and I have limited free time to contribute to clangd.

I get a higher volume of review requests than I'm able to keep up with. I will try to get to this, but I can't promise any particular time frame.

tamird commented:
Hello! Gentle nudge to have a look at this please!

HighCommander4 commented:
Apologies for the long turnaround time. I've had a look at the patch and the changes look reasonable to me, thank you for putting it together.

I just triggered CI and there is a [failure](https://github.com/clangd/node-clangd/actions/runs/13250723393/job/37520404427?pr=32), do you understand the reason for it?

tamird commented:
I do now! I had update node-fetch to 3, which moves to ESM, but there's still a transitive dependency on node-fetch 2, so that's what was being resolved on my local machine. In CI though, the ESM version was being resolved, causing the ESM-related error we saw in CI.

I've downgraded node-fetch to 2 (which is still supported). Things should be green now.

HighCommander4 reviewed (approved):
Thanks! LGTM.

HighCommander4 commented:
I've merged the patch, and the CI workflow is green, however the "publish" workflow [failed](https://github.com/clangd/node-clangd/actions/runs/13447941915/job/37577136689) with:

> npm ERR! Cannot read property 'adm-zip' of undefined

tamird commented:
> I've merged the patch, and the CI workflow is green, however the "publish" workflow [failed](https://github.com/clangd/node-clangd/actions/runs/13447941915/job/37577136689) with:
> 
> > npm ERR! Cannot read property 'adm-zip' of undefined

That workflow is using a very old version of Node: https://github.com/clangd/node-clangd/blob/3f8afdb77c8f46aa3f378d061d42bedfe8b8fbc1/.github/workflows/publish.yml#L10

Do you know why?

HighCommander4 commented:
> > I've merged the patch, and the CI workflow is green, however the "publish" workflow [failed](https://github.com/clangd/node-clangd/actions/runs/13447941915/job/37577136689) with:
> > > npm ERR! Cannot read property 'adm-zip' of undefined
> 
> That workflow is using a very old version of Node:
> 
> https://github.com/clangd/node-clangd/blob/3f8afdb77c8f46aa3f378d061d42bedfe8b8fbc1/.github/workflows/publish.yml#L10
> 
> Do you know why?

I suspect that's the node version that was in common use at the time it was written and no one's updated it since. Would you like to send a PR to bump it to a more modern version?

tamird commented:
https://github.com/clangd/node-clangd/pull/33

HighCommander4 commented:
Thanks -- publishing was now successfully and the new version is up at https://www.npmjs.com/package/@clangd/install
